internal/cli/command/controlplane/debug.go

```diff
@@ -0,0 +1,368 @@
+	logger       logger
+	directories  map[string]bool
+	baseDir      string
+	cwd          string
+	dirPerm      int64
+	filePerm     int64
+	time         func() time.Time
+}
+
+func newTarManager(tarWriter *tar.Writer, errorHandler errorHandler, logger logger, baseDir string) tarManager {
+	tm := tarManager{
+		tarWriter:    tarWriter,
+		errorHandler: errorHandler,
+		logger:       logger,
+		directories:  make(map[string]bool),
+		baseDir:      baseDir,
+		dirPerm:      0777,
+		filePerm:     0666,
+		time:         time.Now,
+	}
+	tm.Cd("/")
+	return tm
+}
+
+// Cd changes the working directory to the given folder in the tar archive.
+// Directories are created implicitly. You can give relative path to the
+// current working directory, or an absolute path starting with a /
+// (where / refers to the baseDir).
+func (t *tarManager) Cd(dir string) {
+	path := t.path(dir)
+	t.Mkdir(path)
+	t.cwd = path
+}
+
+func (t *tarManager) path(dir string) string {
+	if filepath.IsAbs(dir) {
+		return filepath.Clean(filepath.Join("/", dir))
+	}
+	return filepath.Clean(filepath.Join(t.cwd, dir))
+}
+
+// Mkdir creates the given directory and its parents when any of them are missing
+func (t *tarManager) Mkdir(dir string) {
+	path := t.path(dir)
+	if t.directories[path] {
+		return
+	}
+	parent := filepath.Clean(filepath.Join(dir, ".."))
+	if parent != "/" {
+		t.Mkdir(parent)
+	}
+	t.addDir(path)
+	t.directories[path] = true
+}
+
+func (t tarManager) addDir(name string) {
+	// trailing slash makes this a directory, relative paths are normally preferred in the archive
+	name = strings.TrimLeft(filepath.Clean(filepath.Join("/", t.baseDir, name))+"/", "/")
+	err := t.tarWriter.WriteHeader(&tar.Header{Name: name, Mode: t.dirPerm, ModTime: t.time(), ChangeTime: t.time()})
+	if err != nil {
+		t.errorHandler.Handle(errors.WrapIf(err, "failed to create directory in archive"))
+	} else {
+		t.logger.Debugf("added directory %q", name)
+	}
+}
+
+func (t tarManager) addFile(name, content string) (err error) {
+	// relative paths are normally preferred in the archive
+	name = strings.TrimLeft(filepath.Clean(filepath.Join("/", t.baseDir, t.path(name))), "/")
+	bytes := []byte(content)
+	err = t.tarWriter.WriteHeader(&tar.Header{Name: name, Mode: t.filePerm, ModTime: t.time(), ChangeTime: t.time(), Size: int64(len(bytes))})
+	if err != nil {
+		return err
+	}
+
+	n, err := t.tarWriter.Write(bytes)
+	t.logger.Debugf("%d bytes written to %q", n, name)
+	return err
+}
+
+// AddFile adds the file with the given name and content to the archive. The
+// name is relative to the working directory, but it can also be an
+// absolute path (starting from the base directory).
+// Directories are created implicitly.
+func (t *tarManager) AddFile(name string, content interface{}) {
+	path := t.path(name)
+	dir, _ := filepath.Split(path)
+	t.Mkdir(dir)
+
+	var str string
+	switch v := content.(type) {
+	case string:
+		str = v
+	case []byte:
+		str = string(v)
```

Comment:
afaik it can't cause a problem in this specific case, because it's casted back to []byte before writing, and casting `[]byte` and `string` back and forth itself doesn't change the content or the storage format. if the input is neither a string or a byte slice, the String method is supposed to output proper (utf8) strings.

Reply:
Yeah, I didn't connect the cast back, so then it is fine.